libraries/package-extractor/src/AssetHandler.ts

```diff
@@ -0,0 +1,224 @@
+  private readonly _archiveManager: ArchiveManager | undefined;
+  private readonly _archiveFilePath: string | undefined;
+  private readonly _linkCreationMode: LinkCreationMode;
+  private readonly _includedAssetPaths: Set<string> = new Set<string>();
+  private _isFinalized: boolean = false;
+
+  public constructor(options: IAssetHandlerOptions) {
+    const {
+      terminal,
+      sourceRootFolder,
+      targetRootFolder,
+      linkCreation,
+      symlinkAnalyzer,
+      createArchiveFilePath,
+      createArchiveOnly = false
+    } = options;
+    this._terminal = terminal;
+    this._sourceRootFolder = sourceRootFolder;
+    this._targetRootFolder = targetRootFolder;
+    this._symlinkAnalyzer = symlinkAnalyzer;
+    if (createArchiveFilePath) {
+      if (path.extname(createArchiveFilePath) !== '.zip') {
+        throw new Error('Only archives with the .zip file extension are currently supported.');
+      }
+      this._archiveFilePath = path.resolve(targetRootFolder, createArchiveFilePath);
+      this._archiveManager = new ArchiveManager();
+    }
+    if (createArchiveOnly && !this._archiveManager) {
+      throw new Error('createArchiveOnly cannot be true if createArchiveFilePath is not provided');
+    }
+    this._createArchiveOnly = createArchiveOnly;
+    this._linkCreationMode = linkCreation || 'default';
+  }
+
+  public async includeAssetAsync(options: IIncludeAssetPathOptions): Promise<void>;
+  public async includeAssetAsync(options: IIncludeExistingAssetPathOptions): Promise<void>;
+  public async includeAssetAsync(options: IIncludeAssetContentOptions): Promise<void>;
+  public async includeAssetAsync(options: IIncludeAssetOptions): Promise<void> {
+    const { sourceFileContent, targetFilePath, ignoreIfExisting = false } = options;
+    let { sourceFilePath } = options;
+
+    if (this._isFinalized) {
+      throw new Error('includeAssetAsync() cannot be called after finalizeAsync()');
+    }
+    if (!sourceFilePath && !sourceFileContent) {
+      if (!Path.isUnder(targetFilePath, this._targetRootFolder)) {
+        throw new Error('The existing asset path must be under the target root folder');
+      }
+      sourceFilePath = targetFilePath;
+    }
+    if (sourceFilePath && sourceFileContent) {
+      throw new Error('Either sourceFilePath or sourceFileContent must be provided, but not both');
+    }
+    if (this._includedAssetPaths.has(targetFilePath)) {
+      if (ignoreIfExisting) {
+        return;
+      }
+      throw new Error(`The asset at path "${targetFilePath}" has already been included`);
+    }
+
+    if (!this._createArchiveOnly) {
+      // Ignore when the source file is the same as the target file, as it's a no-op
+      if (sourceFilePath && sourceFilePath !== targetFilePath) {
+        // Use the fs.copyFile API instead of FileSystem.copyFileAsync() since copyFileAsync performs
+        // a needless stat() call to determine if it's a file or folder, and we already know it's a file.
```

Comment:
Can probably update this to try to copy as a file, and then fall back to copying as a folder if that fails.

Reply:
Idk, I actually think if the target is a directory that should fail. The alternate source option (`sourceFileContent`) can only ever be a file. Seems appropriate for the `EISDIR` error to be thrown